spec/index.bs

```diff
@@ -2036,12 +2043,150 @@ dictionary GPUShaderModuleDescriptor : GPUObjectDescriptorBase {
 
+## Base pipelines # {#pipeline-base}
+
 <script type=idl>
 dictionary GPUPipelineDescriptorBase : GPUObjectDescriptorBase {
-    required GPUPipelineLayout layout;
+    GPUPipelineLayout layout;
+};
+
+interface mixin GPUPipelineBase {
+    GPUBindGroupLayout getBindGroupLayout(unsigned long index);
 };
 </script>
 
+{{GPUPipelineBase}} has the following internal slots:
+<dl dfn-type=attribute dfn-for="GPUPipelineBase">
+    : <dfn>\[[layout]]</dfn> of type `GPUPipelineLayout`.
+    ::
+        The definition of the layout of resources which can be used with `this`.
+</dl>
+
+### <dfn method for=GPUPipelineBase>getBindGroupLayout(index)</dfn> ### {#pipelinebase-getBindGroupLayout}
+
+<div algorithm="GPUPipelineBase.getBindGroupLayout">
+    **Arguments:**
+
+        - {{unsigned long}} |index|
+
+    **Returns:** {{GPUBindGroupLayout}}
+
+    1. If |index| is greater or equal to {{GPULimits/maxBindGroups}}:
+
+        1. Throw a {{RangeError}}.
+
+    1. If |this| is not [=valid=]:
+
+        1. Return a new error {{GPUBindGroupLayout}}.
+
+    1. Return a new {{GPUBindGroupLayout}} wrapping
+        |this|.{{GPUPipelineBase/[[layout]]}}.{{GPUPipelineLayout/[[bindGroupLayout]]}}[|index|].
+
+    Note: Only returning new {{GPUBindGroupLayout}} objects ensure no synchronization is necessary
+        between the [=Content timeline=] and the [=Device timeline=].
+
+</div>
+
+### Default pipeline layout ### {#default-pipeline-layout}
+
+A {{GPUPipelineBase}} object that was created without a {{GPUPipelineDescriptorBase/layout}}
+has a default layout created and used instead.
+
+<div algorithm="default pipeline layout creation">
+
+    1. Let |groupDescs| be a sequence of {{GPULimits/maxBindGroups}} new {{GPUBindGroupLayoutDescriptor}} objects.
+    1. For each |groupDesc| in |groupDescs|:
+
+        1. Set |groupDesc|.{{GPUBindGroupLayoutDescriptor/entries}} to an empty sequence.
+
+    1. For each {{GPUProgrammableStageDescriptor}} |stageDesc| in the descriptor used to create the pipeline:
+
+        1. Let |stageInfo| be the "reflection information" for |stageDesc|.
+        1. Let |shaderStage| be the {{GPUShaderStageFlags}} for |stageDesc|.{{GPUProgrammableStageDescriptor/entryPoint}}
+            in |stageDesc|.{{GPUProgrammableStageDescriptor/module}}.
+        1. For each resource |resource| in |stageInfo|'s resource interface.
+
+            1. Let |group| be |resource|'s "group" decoration.
+            1. Let |binding| be |resource's "binding" decoration.
+            1. Let |entry| be a new {{GPUBindGroupLayoutEntry}}.
+            1. Set |entry|.{{GPUBindGroupLayoutEntry/binding}} to |binding|.
+            1. Set |entry|.{{GPUBindGroupLayoutEntry/visibility}} to |shaderStage|.
+            1. If |resource| is for a sampler binding:
+
+                    1. Set |entry|.{{GPUBindGroupLayoutEntry/type}} to
+                        {{GPUBindingType/sampler}}.
+
+            1. If |resource| is for a comparison sampler binding:
```

Comment:
How is the shader reflection supposed to know if this is a comparison sampler binding?

Reply:
My understanding is that SPIRV doesn't surface this in the type, but SPIRV-Cross does static analysis to decide how the sampler is used. Probably something we should enforce properly in WGSL. Added a note about this in https://github.com/gpuweb/gpuweb/issues/574

Reply:
It's a good candidate to add to WGSL semantics. Still a bit worrying that SPIR-V doesn't have it and requires extra work :/